preupgradechecks/preupgradechecks.go

```diff
@@ -0,0 +1,207 @@
+			break
+		}
+	}
+
+	if err != nil {
+		fatal(fmt.Sprintf("Error: %v listing functions even after %d retries", err, MaxRetries))
+	}
+
+	// check that all secrets, configmaps, packages are in the same namespace
+	for _, fn := range fList.Items {
+		secrets := fn.Spec.Secrets
+		for _, secret := range secrets {
+			if secret.Namespace != fn.Metadata.Namespace {
+				result = multierror.Append(result, fmt.Errorf("function : %s.%s cannot reference a secret : %s in namespace : %s", fn.Metadata.Name, fn.Metadata.Namespace, secret.Name, secret.Namespace))
+			}
+		}
+
+		configmaps := fn.Spec.ConfigMaps
+		for _, configmap := range configmaps {
+			if configmap.Namespace != fn.Metadata.Namespace {
+				result = multierror.Append(result, fmt.Errorf("function : %s.%s cannot reference a configmap : %s in namespace : %s", fn.Metadata.Name, fn.Metadata.Namespace, configmap.Name, configmap.Namespace))
+			}
+		}
+
+		if fn.Spec.Package.PackageRef.Namespace != fn.Metadata.Namespace {
+			result = multierror.Append(result, fmt.Errorf("function : %s.%s cannot reference a package : %s in namespace : %s", fn.Metadata.Name, fn.Metadata.Namespace, fn.Spec.Package.PackageRef.Name, fn.Spec.Package.PackageRef.Namespace))
+		}
+	}
+
+	if result != nil {
+		log.Printf("Installation failed due to the following errors :")
+		log.Printf("Summary : A function cannot reference secrets, configmaps and packages outside it's own namespace")
+		fatal(result.Error())
+	}
+
+	log.Printf("Function Spec References verified")
+}
+
+// deleteClusterRoleBinding deletes the clusterRoleBinding passed as an argument to it.
+// if its not present, it just ignores and returns no errors
+func (client *Client) deleteClusterRoleBinding(clusterRoleBinding string) (err error) {
+	for i := 0; i < MaxRetries; i++ {
+		err = client.k8sClient.RbacV1beta1().ClusterRoleBindings().Delete(clusterRoleBinding, &metav1.DeleteOptions{})
+		if err != nil && k8serrors.IsNotFound(err) || err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+// RemoveClusterAdminRolesForFissionSAs deletes the clusterRoleBindings previously created on this cluster
+func (client *Client) RemoveClusterAdminRolesForFissionSAs() {
+	clusterRoleBindings := []string{"fission-builder-crd", "fission-fetcher-crd"}
+	for _, clusterRoleBinding := range clusterRoleBindings {
+		err := client.deleteClusterRoleBinding(clusterRoleBinding)
+		if err != nil {
+			fatal(fmt.Sprintf("Error deleting rolebinding : %s, err : %v", clusterRoleBinding, err))
+		}
+	}
+
+	log.Println("Removed cluster admin privileges for fission-builder and fission-fetcher Service Accounts")
+}
+
+// NeedRoleBindings checks if there is atleast one package or function in default namespace.
+// this is needed to find out if package-getter-rb and secret-configmap-getter-rb needs to be created for fission-fetcher
+// and fission-builder service accounts.
+// this because, we just deleted the ClusterRoleBindings for these service accounts in the previous function and
```

Comment:
This is* because.  Also, please capitalize the first letters of sentences.  Makes long comments more readable.